File: instagram4a_/src/main/java/com/bodiart/instagram4a/requests/base/InstagramPostRequest.java
package com.bodiart.instagram4a.requests.base;


import com.bodiart.instagram4a.InstagramConstants;
import com.bodiart.instagram4a.util.InstagramHashUtil;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * Created by root on 08/06/17.
 */

public abstract class InstagramPostRequest<T> extends InstagramRequest<T> {

    @Override
    public String getMethod() {
        return "POST";
    }

    @Override
    public T execute() throws IOException, InstantiationException, IllegalAccessException {

        Request request = new Request.Builder()
                .url(InstagramConstants.API_URL + getUrl())
                .addHeader("Connection", "close")
                .addHeader("Accept", "*/*")
                .addHeader("Cookie2", "$Version=1")
                .addHeader("Accept-Language", "en-US")
                .addHeader("X-IG-Capabilities", "3boBAA==")
                .addHeader("X-IG-Connection-Type", "WIFI")
                .addHeader("X-IG-Connection-Speed", "-1kbps")
                .addHeader("X-IG-App-ID", "567067343352427")
                .addHeader("User-Agent", InstagramConstants.USER_AGENT)
                .post(RequestBody.create(MediaType.parse("application/x-www-form-urlencoded"), InstagramHashUtil.generateSignature(getPayload())))
                .build();

        Response response = api.getClient().newCall(request).execute();
        api.setLastResponse(response);

        int resultCode = response.code();
        String content = response.body().string();

        return parseResult(resultCode, content);
    }
}

File: instagram4a_/src/main/java/com/bodiart/instagram4a/refactor/kotlin/request/base/InstagramPostRequest.kt
package com.bodiart.instagram4a.refactor.kotlin.request.base

import com.bodiart.instagram4a.refactor.kotlin.utils.InstagramConstants
import com.bodiart.instagram4a.refactor.kotlin.utils.InstagramHashUtil
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody

abstract class InstagramPostRequest <T> : InstagramRequest<T>() {

    override fun getMethod() = "POST"

    override fun execute(): T {

        var baseUrl = InstagramConstants.API_URL
        if (requireApiV2())
            baseUrl = InstagramConstants.API_URL2

        val request = Request.Builder()
                .url(baseUrl + getUrl())
                .addHeader("Connection", "close")
                .addHeader("Accept", "*/*")
                .addHeader("Cookie2", "\$Version=1")
                .addHeader("Accept-Language", "en-US")
                .addHeader("X-IG-Capabilities", "3boBAA==")
                .addHeader("X-IG-Connection-Type", "WIFI")
                .addHeader("X-IG-Connection-Speed", "-1kbps")
                .addHeader("X-IG-App-ID", "567067343352427")
                .addHeader("User-Agent", InstagramConstants.USER_AGENT)
                .post(RequestBody.create("application/x-www-form-urlencoded".toMediaTypeOrNull(),
                        InstagramHashUtil.generateSignature(getPayload())))
                .build()

        val response = api.client.newCall(request).execute()

        return parseResult(response.code, response.body?.string())
    }
}